xarray/core/coordinates.py

```diff
@@ -796,6 +789,113 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(Coordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+
+    This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    This will add both the coordinates variables and their index.
+    """
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    _data: DataTree  # type: ignore[assignment]  # complaining that DataTree is not a subclass of DataWithCoords - this can be fixed by refactoring, see #9203
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
```

Comment:
maybe worth updating ;)

Reply:
updated in https://github.com/pydata/xarray/pull/9451/commits/af94af4a68b0d7b2181f47fd0ba266ecb575cb3c